sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -60,17 +60,28 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
 
   override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
     val javaType = ctx.javaType(dataType)
-    val value = ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString)
+    val value = if (!ctx.isColumnarType(ctx.INPUT_ROW)) {
+      ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString)
+    } else {
+      ctx.getValue(ctx.INPUT_ROW, dataType, ctx.INPUT_COLORDINAL)
```

Comment:
Is it good to introduce new ```InputReference```, which is similar to ```BoundReference```, only for the code to access ```ColumnBatch```?